### PR TITLE
Make EC2 Discovery Plugin Retry Requests

### LINKED ISSUE #50462: [discovery-ec2] Plugin doesn't retry EC2 describe instance calls
### Problem Description
Discovery EC2 doesn't retry EC2 describe calls on throttle or exceptions due to `NO_RETRY_CONDITION`  which has the potential to make the throttling worse. Although there is a logic to exponentially backoff in such cases, backoff doesn't kick in.

https://github.com/elastic/elasticsearch/blob/7203ceefd8761cdbfd9d4510cdda1ac968caf11d/plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java#L82-L90

AWS [reference](https://github.com/aws/aws-sdk-java/blob/de3ea40fbd783f1a667f24303399f5ed2a0279ec/aws-java-sdk-core/src/main/java/com/amazonaws/retry/RetryPolicy.java#L140-L148)

### PR DESCRIPTION
**Pull request by original-brownbear**
Use the default retry condition instead of never retrying in the discovery plugin causing hot retries upstream and
add a test that verifies retrying works.

Closes #50462

### CODE REVIEW

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+
+@SuppressForbidden(reason = "use a http server")
+public class EC2RetriesTests extends ESTestCase {
```
Comment (original-brownbear) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java]: Admittedly quite a bit of code, but I'm hoping we can just extend the REST layer mocking here and dry things up against the fixture step-by-step like we did for S3 and the other cloud providers in the repository tests to get rid of all our complicated mocking of actual AWS SDK classes and just mock on the REST layer.

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java [hunk 2]
```diff
@@ -79,7 +79,7 @@ static ClientConfiguration buildConfiguration(Logger logger, Ec2ClientSettings c
         // Increase the number of retries in case of 5xx API responses
         final Random rand = Randomness.get();
         final RetryPolicy retryPolicy = new RetryPolicy(
-            RetryPolicy.RetryCondition.NO_RETRY_CONDITION,
+            null,
```
Comment (original-brownbear) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java]: Just using the default retry condition here now. We may want to make the retry interval configurable, if only internally for tests. The current settings make the first retry happen after almost 20s but I didn't want to add that complication here yet.

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java [hunk 3]
```diff
@@ -79,7 +79,7 @@ static ClientConfiguration buildConfiguration(Logger logger, Ec2ClientSettings c
         // Increase the number of retries in case of 5xx API responses
         final Random rand = Randomness.get();
         final RetryPolicy retryPolicy = new RetryPolicy(
```
Comment (DaveCTurner) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java]: Can we avoid setting our own retry policy at all and use the SDK default? The SDK makes a more nuanced decision about how much to back off based on the error type. If we really want 10 retries then that can be overridden with `setMaxErrorRetry(10)` instead?

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I left a question.

I also think we should adjust the behaviour of `AwsEc2SeedHostsProvider` to respect `discovery.ec2.node_cache_time` even if the last attempt resulted in an exception and/or an empty list.

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java (hunk 3 above)
Comment (original-brownbear) [reply plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java]: Right, that's much nicer. Moved to the default and kept retries at `10` for now in https://github.com/elastic/elasticsearch/pull/50550/commits/d015576900c80123733e36ce2a8f5a4b8c897b41 :)

**Comment by original-brownbear** [comment]
> I also think we should adjust the behaviour of AwsEc2SeedHostsProvider to respect discovery.ec2.node_cache_time even if the last attempt resulted in an exception and/or an empty list.

Wanna do that here or in a follow-up?

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+                            responseBody = generateDescribeInstancesResponse(hosts);
+                            break;
+                        }
+                    }
+                    responseBody = responseBody == null ? new byte[0] : responseBody;
+                    exchange.getResponseHeaders().set("Content-Type", "text/xml; charset=UTF-8");
+                    exchange.sendResponseHeaders(HttpStatus.SC_OK, responseBody.length);
+                    exchange.getResponseBody().write(responseBody);
+                }
+            }
```
Comment (DaveCTurner) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java]:
```suggestion
            }

            fail("did not send response");
```

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                    for (NameValuePair parse : URLEncodedUtils.parse(request, UTF_8)) {
+                        if ("Action".equals(parse.getName())) {
+                            responseBody = generateDescribeInstancesResponse(hosts);
+                            break;
+                        }
+                    }
+                    responseBody = responseBody == null ? new byte[0] : responseBody;
+                    exchange.getResponseHeaders().set("Content-Type", "text/xml; charset=UTF-8");
+                    exchange.sendResponseHeaders(HttpStatus.SC_OK, responseBody.length);
+                    exchange.getResponseBody().write(responseBody);
```
Comment (DaveCTurner) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java]:
```suggestion
                    exchange.getResponseBody().write(responseBody);
                    return;
```

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        httpServer.createContext("/", exchange -> {
+            if (exchange.getRequestMethod().equals(HttpMethodName.POST.name())) {
+                final String request = new String(exchange.getRequestBody().readAllBytes(), UTF_8);
+                final String userAgent = exchange.getRequestHeaders().getFirst("User-Agent");
+                if (userAgent != null && userAgent.startsWith("aws-sdk-java")) {
+                    final String auth = exchange.getRequestHeaders().getFirst("Authorization");
+                    if (auth == null || auth.contains(accessKey) == false) {
+                        throw new IllegalArgumentException("wrong access key: " + auth);
+                    }
+                    if (failedRequests.add(exchange.getRequestHeaders().getFirst("Amz-sdk-invocation-id"))) {
```
Comment (DaveCTurner) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java]: Could we also check that we retry repeatedly? The first five retries take less than a second.

**Comment by DaveCTurner** [comment]
> Wanna do that here or in a follow-up?

Yes a follow-up makes more sense.

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java (hunk 6 above)
Comment (original-brownbear) [reply plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java]: Done :)

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java (hunk 5 above)
Comment (original-brownbear) [reply plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/EC2RetriesTests.java]: Makes sense applied together with the otehr suggestion :)

**Comment by original-brownbear** [comment]
@DaveCTurner thanks, all points addressed I think :)

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**